In [ ]:
# 1 - importação de bibliotecas:
import pandas as pd # Módulo para a leitura da base de dados
#import plotly.express as px # Módulo para a representaçao gráfica da base de dados

In [3]:
# 2 - ler a base de dados:
tabela = pd.read_csv('retail_sales_dataset.csv')

In [4]:
# 2.1 - imprimir a base de dados:
display(tabela)

,Transaction ID,Date,Customer ID,Gender,Age,Product Category,Quantity,Price per Unit,Total Amount
0,1,2023-11-24,CUST001,Male,34,Beauty,3,50,150
1,2,2023-02-27,CUST002,Female,26,Clothing,2,500,1000
2,3,2023-01-13,CUST003,Male,50,Electronics,1,30,30
3,4,2023-05-21,CUST004,Male,37,Clothing,1,500,500
4,5,2023-05-06,CUST005,Male,30,Beauty,2,50,100
...,...,...,...,...,...,...,...,...,...
995,996,2023-05-16,CUST996,Male,62,Clothing,1,50,50
996,997,2023-11-17,CUST997,Male,52,Beauty,3,30,90
997,998,2023-10-29,CUST998,Female,23,Beauty,4,25,100
998,999,2023-12-05,CUST999,Female,36,Electronics,3,50,150


In [5]:
# 2.2 - Imprimir as informações da tabela
tabela.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   Transaction ID    1000 non-null   int64
 1   Date              1000 non-null   str  
 2   Customer ID       1000 non-null   str  
 3   Gender            1000 non-null   str  
 4   Age               1000 non-null   int64
 5   Product Category  1000 non-null   str  
 6   Quantity          1000 non-null   int64
 7   Price per Unit    1000 non-null   int64
 8   Total Amount      1000 non-null   int64
dtypes: int64(5), str(4)
memory usage: 70.4 KB


Observações do tratamento:

1) A tabela possui duas informações irrelevantes para a análise exploratória de dados: Transaction ID e Customer ID.
2) A tabela possui uma variável com um tipo de dado que excede a sua necessidade de espaço na memória: Age.

In [6]:
# 3 - Tratar a base de dados
for coluna in tabela.columns: # Adicionando _ no local de espaço em branco e modificando as letras das colunas para minúsculo.

    coluna_nova = coluna.split()
    coluna_nova = '_'.join(coluna_nova)

    coluna_nova = coluna_nova.lower()

    tabela = tabela.rename(columns={coluna: coluna_nova})

tabela = tabela.drop(columns=['transaction_id', 'customer_id']) # Descartando as colunas desnecessárias

tabela['age'] = tabela['age'].astype('int8') # Alterando o tipo de dado int64 para int8

# 3.1 - imprimir a tabela depois de modificada:
display(tabela)
tabela.info()

,date,gender,age,product_category,quantity,price_per_unit,total_amount
0,2023-11-24,Male,34,Beauty,3,50,150
1,2023-02-27,Female,26,Clothing,2,500,1000
2,2023-01-13,Male,50,Electronics,1,30,30
3,2023-05-21,Male,37,Clothing,1,500,500
4,2023-05-06,Male,30,Beauty,2,50,100
...,...,...,...,...,...,...,...
995,2023-05-16,Male,62,Clothing,1,50,50
996,2023-11-17,Male,52,Beauty,3,30,90
997,2023-10-29,Female,23,Beauty,4,25,100
998,2023-12-05,Female,36,Electronics,3,50,150


<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   date              1000 non-null   str  
 1   gender            1000 non-null   str  
 2   age               1000 non-null   int8 
 3   product_category  1000 non-null   str  
 4   quantity          1000 non-null   int64
 5   price_per_unit    1000 non-null   int64
 6   total_amount      1000 non-null   int64
dtypes: int64(3), int8(1), str(3)
memory usage: 48.0 KB


Conclusões de tratamento

1) Redução do tempo de percurso pela tabela
2) Redução da ocupação de memória da tabela em aproximadamente 31,82% (70.4 -> 48.0), que pode ser um valor alto para tabelas muito grandes


In [7]:
# 4 - Verificar a relevância dos dados
tabela['product_category'].value_counts()

product_category
Clothing       351
Electronics    342
Beauty         307
Name: count, dtype: int64

In [8]:
tabela['product_category'].value_counts(normalize=True)


product_category
Clothing       0.351
Electronics    0.342
Beauty         0.307
Name: proportion, dtype: float64

Ao examinar essa distribuição, percebe-se que há a tendência da empresa vender mais produtos da categoria "Clothing" e "Electronics" no período de exercício, porém a categoria "Beauty" não é irrelevante a ponto de se planejar previsões ou mobilizações substanciais.

In [9]:
tabela['age'].value_counts() # Descobrindo a distribuição de idade

age
64    31
43    31
51    30
57    30
34    28
54    28
22    27
62    27
42    26
47    26
46    25
23    24
40    24
50    23
27    23
26    22
30    22
52    22
35    22
31    22
60    22
19    21
18    21
28    21
55    21
20    21
41    21
21    20
25    20
49    19
38    19
56    19
32    19
39    18
48    18
61    18
45    17
63    17
59    17
53    17
37    16
29    16
44    15
36    15
24    15
58    14
33    10
Name: count, dtype: int64

In [ ]:
# Análise: Percentual da faixa etária
percentual_idosos = (tabela['age'] >= 50).mean() * 100 # (tabela['age'] >= 50) -> acumulador de True, que é ni .mean() -> ni / n (distribuição relativa)
percentual_jovens = ((tabela['age'] >= 15) & (tabela['age'] <= 24)).mean() * 100
percentual_jovens_adultos = ((tabela['age'] > 24) & (tabela['age'] < 50)).mean() * 100
print('percentual de idade dos clientes: \n') 
print(percentual_idosos)
print(percentual_jovens_adultos)
print(percentual_jovens)


percentual de idade dos clientes: 

33.6
51.5
14.899999999999999


Conclusão: Nota-se que mais da metade dos clientes estão entre 24 e 50 anos de idade 

In [11]:
# Distribuição do sexo biológico
tabela['gender'].value_counts()

gender
Female    510
Male      490
Name: count, dtype: int64

In [21]:
# Distribuição de data de compra
ni_data = tabela['date'].value_counts()
print(ni_data)

date
2023-05-16    11
2023-07-14    10
2023-05-23     9
2023-08-05     8
2023-02-05     8
              ..
2023-03-02     1
2023-08-02     1
2023-04-17     1
2023-03-30     1
2023-05-28     1
Name: count, Length: 345, dtype: int64


In [ ]:
# Percentual da quantidade de compras que há em uma data específica
percentual_7_11 = ((tabela['date'].value_counts() >= 7) & (tabela['date'].value_counts() <= 11)).mean() * 100
percentual_3 = (tabela['date'].value_counts() == 3).mean() * 100

print(f'mediana: {tabela['date'].value_counts().median()}')
print(f'média: {tabela['date'].value_counts().mean()}') #\n\n\n\n
print(f'Percentual da distribuição de 7 a 11: {percentual_7_11}')
print(f'percentual da distribuição igual a 3: {percentual_3}')

mediana: 3.0
média: 2.898550724637681
Desvio padrão dos meses com frequência absoluta maior ou igual a 7: 18.099723754798028

Percentual da distribuição de 7 a 11: 2.0289855072463765
percentual da distribuição igual a 3: 27.82608695652174


Observa-se, pela mediana, que a empresa atende, em sua maioria, aproximadamente 3 clientes por dia, tendo a uma frequência relativa (Fi) 13.8, aproximadamente, vezes maior que a frequência relativa acumulada (Fia) do intervalo discreto 7 a 11.

Portanto, vê-se uma assimetria que representa um aumento significativo de vendas em determinadas datas. Como somente os dias ditam os momentos de ápice de vendas, infere-se que haja talvez a falta de organização ou de conhecimento sobre o momento correto de ceder descontos ou semelhantes aos clientes.

In [14]:
# Distribuição da quantidade de compras por cliente
tabela['quantity'].value_counts()

quantity
4    263
1    253
2    243
3    241
Name: count, dtype: int64

In [27]:
# 4.1 Medidas-resumo:
tabela.describe()

,age,quantity,price_per_unit,total_amount
count,1000.00000,1000.000000,1000.000000,1000.000000
mean,41.39200,2.514000,179.890000,456.000000
std,13.68143,1.132734,189.681356,559.997632
min,18.00000,1.000000,25.000000,25.000000
25%,29.00000,1.000000,30.000000,60.000000
50%,42.00000,3.000000,50.000000,135.000000
75%,53.00000,4.000000,300.000000,900.000000
max,64.00000,4.000000,500.000000,2000.000000
